100_Numpy_exercises_with_hints_with_solutions.md

```diff
@@ -673,10 +673,10 @@ print(F)
 # Author: Nadav Horesh
 
 w,h = 16,16
-I = np.random.randint(0,2,(h,w,3)).astype(np.ubyte)
+I = np.random.randint(0,2,(h,w,3))
```

Comment:
Why did you remove the cast to ubyte ?

Reply:
In a sufficiently long array, the expected number of different "colors" should be `2**3 == 8`. For the solution to work, `F` must be a matrix of at least 24-bit integers, but this is not what happens on my machine: `I[..., 1] * 256` has type `np.uint16`, and the same happens with `I[..., 0] * 256 * 256`, leading to overflow. Writing `(I[..., 0] * 256) * 256` doesn't solve the problem. A cast to a sufficient wide type has to happen at some point, or the initial cast has to go.

```
>>> import numpy as np
>>> w, h = 16, 16
>>> I = np.random.randint(0, 2, (h, w, 3))
>>> I2 = I.astype(np.ubyte)
>>> F = I[...,0]*256*256 + I[...,1]*256 +I[...,2]
>>> F2 = I2[...,0]*256*256 + I2[...,1]*256 +I2[...,2]
>>> print(len(np.unique(F)), len(np.unique(F2)))
8 4
>>> print(F.dtype, F2.dtype)
int64 uint16
```

Reply:
ok